### ISSUE TITLE
add @hierarchy slot and methods

### ISSUE THREAD

**Issue by zkamvar** [thibautjombart/adegenet#16]
These will come from poppr

**Comment by thibautjombart** [comment]
Looks like the check is passing with this now. Lemme know when you think stuff is release-ready.

**Comment by zkamvar** [comment]
I will. I still need to add hierarchy methods and tests for the following:

 - [ ] `rbind.genlight()`
 - [ ] `seppop()`
 - [ ] `seploc()`
 - [ ] `repool()`

As a side note, @jgx65 mentioned earlier that hierfstat currently spends a lot of time validating that the hierarchies are indeed hierarchical. Currently (and ironically), the hierarchy slot can hold hierarchies, but currently does not validate that they are so. If @jgx65 can provide a general function to validate that hierarchical levels provided are hierarchies, then we could add a small slot in the objects that will give the results of the validation by a hierarchical formula. This would allow hierfstat's functions to run more efficiently by simply checking if the user input for the hierarchies follow the order provided in the slot. What do you think?

**Comment by thibautjombart** [comment]
I think it makes perfect sense, and is part of using S4 classes: as long as
you use these objects, you should be guaranteed that their content is what
you expect. Ideally the checks should only be done:

1) when creating the object using "new(...)"
2) when calling @[slotName] <- [newValue]

Note that in adegenet, in multiple places I call validObject(...) at the
beginning of functions.. I'm not sure this is actually useful any more.

**Comment by zkamvar** [comment]
Along the same lines of the checks being performed only once:

I was thinking of having the hierarchy slot default to being unchecked since factorial sampling designs are also useful (especially in my field).

When the user wants to confirm that they have a hierarchical design they could do something like this:

```r
checkhierarchy(object) <- ~pop/subPop/subSubPop
```
This would give an error if these levels are not hierarchical. If they are hierarchical, the formula would be saved in the new slot (possibly name it hiercheck or something goofy like that). This would allow the user to mix hierarchical and factorial sampling, but specify which structures are strictly hierarchical. That way, the user's workflow could be:

```r
checkhierarchy(object) <- ~pop/subPop/subSubPop # assert that this is hierarchical
varcomp.glob(~pop/subPop, object) # test on given hierarchy
```

**Comment by jgx65** [comment]
You're doing a great job guys. I'll look into this later on this afternoon or more probably tomorrow (I can't keep your pace!)


Cheers






Jérôme Goudet
Dept. Ecology & Evolution
Biophore, UNIL-Sorge
CH-1015 Lausanne
http://www.unil.ch/dee/home/menuinst/research/group-goudet.html
mail:jerome.goudet@unil.ch
Tel:+41 (0)21 692 4242

________________________________

**Comment by romunov** [comment]
I was about to say the same thing. I'm barely able to keep up with the reading, let alone coding. :)

**Comment by thibautjombart** [comment]
@zkamvar what you describe makes sense. A tad unusual so we'll need to document that, but we have to anyway ;) 

@zkamvar @jgx65 Out of curiosity: how long does it take to check hierarchy? If ~1s or less, then probably no point in being overly fancy - we can check it whenever needed.

**Comment by zkamvar** [comment]
@thibautjombart @jgx65 I would argue that, regardless of time it takes to check a hierarchy, this should be an automated process so that it can be utilized by other packages that come along in the future.

**Comment by thibautjombart** [comment]
I agree if we have a strict definition for the slot. But if we allow for
the slot to store non hierarchical stuff as well then testing can/should be
externalised as an exported function.
Makes sense ?

**Comment by zkamvar** [comment]
That makes perfect sense.

I think we might be saying the same thing, just in slightly different ways.

I am thinking we should have two separate slots, one to store strata (including hierarchies) (as a data.frame) and another specifically to define which components of the strata are hierarchical (as a formula). 

In my branch, I currently have the data frame in the "hierarchy" slot, but it can store strata as well. To avoid confusion, I can move the data frame to a new slot called "strata" and add the formula that references the columns of the "strata" slot to the "hierarchy" slot. This does affect poppr, as I have defined methods like `gethierarchy` and `addhierarchy`, but I'm already making a bunch of changes so I don't mind removing those, killing backwards compatibility, and increasing the major version number.

**Warning, long example below**

***

To give a clear example, here's what I can envision for a workflow. Say the user has a genind object and the strata and hierarchy slots are NULL

```r
> getstrata(object)
NULL
> gethierarchy(object)
NULL
```

Let's say the user has a data frame that has a data frame with six columns in it:

|Pop|Subpop|Subsubpop|Year|Phenotype|Age|
|---|---|---|---|---|---|
| ... | ... | ... | ... | ... | ... |

The first three columns are hierarchical, but the last three are not. The user would first set the strata and then define the hierarchies.

```r
setstrata(object) <- df
sethierarchy(object) <- ~Pop/Subpop/Subsubpop
```
The check for hierarchical structure will be performed when `sethierarchy` is called. From there, the accessor `gethierarchy` will simply check the order of the formula provided, which will be much faster than checking a data frame. This makes it so that authors of other packages need only utilize `gethierarchy` to assert that the formula provided by the user is hierarchical. 

Now, when the user uses the accessors:

```r
> class(object@strata)
[1] "data.frame"
> getstrata(object)
```

|Pop|Subpop|Subsubpop|Year|Phenotype|Age|
|---|---|---|---|---|---|
| ... | ... | ... | ... | ... | ... |

```r
> class(object@hierarchy)
[1] "formula"
> gethierarchy(object)
```

|Pop|Subpop|Subsubpop|
|---|---|---|
| ... | ... | ... |

```r
> gethierarchy(object, ~Pop/Subpop)
```

|Pop|Subpop|
|---|---|
| ... | ... |

If the user attempts to get a hierarchy that is not defined, or doesn't make sense, an error will return:

```r
> gethierarchy(object, ~Subpop/Pop)
error: not hierarchical
> gethierarchy(object, ~Pop/Year)
error: Year is not part of the hierarchy
```

Sorry for the length.

**Comment by thibautjombart** [comment]
Haha, don't be sorry for providing a detailed & useful example ;) 
It looks perfect to me. For consistency, and though I'm using get.xxx or getXxx in other packages, we may want to name the accessors 'hierarchy' and 'stata' in adegenet, otherwise looks great to me. I like the idea of using formula supported by a data.frame to store the information. Much more flexible that way, and a lot easier to use for the user as well. 

I can see some minor cool tweaks I'll want to do once this is in place, e.g. looking up this data.frame for 'pop' info, so that genind2genpop(x, "Year") would work in your example. But most importantly, any hierarchical analysis will be readily available through your formula interface, which is really really cool!

**Comment by thibautjombart** [comment]
Merge done. Closing issue.
